File: src/runtime/auxiliary-windows.ts
import { invoke } from "@tauri-apps/api/core";
import { getCurrentWindow } from "@tauri-apps/api/window";

export const AUXILIARY_CONTROLS_LABEL = "auxiliary-screen-sharing-controls";
export const AUXILIARY_STATE_EVENT = "auxiliary-window-state";
export const AUXILIARY_ACTION_EVENT = "auxiliary-window-action";

/** Both the native label and bundled route must match; an auxiliary view never mounts App. */
export function resolveWindowView(label: string, search: string) {
  if (label === "main") return "main";
  if (
    label === AUXILIARY_CONTROLS_LABEL &&
    new URLSearchParams(search).get("auxiliary") === "screen-sharing-controls"
  ) {
    return "screen-sharing-controls";
  }
  return null;
}

export interface ScreenSharingSnapshot {
  readonly revision: string;
  /** Changes with the main owner or marker setting, never with capture progress. */
  readonly pointerRevision: string;
  readonly available: boolean;
  readonly active: boolean;
  readonly busy: boolean;
  readonly pointersEnabled: boolean;
  readonly pointersReady: boolean;
  readonly sources: readonly { readonly id: number; readonly name: string }[];
  readonly sourceId: number | null;
  readonly intervalSeconds: number;
  readonly hasError: boolean;
  readonly lastObservedAt: number | null;
  readonly language: "en" | "ja";
}

export interface PublishedAuxiliarySnapshot {
  readonly version: number;
  readonly snapshot: ScreenSharingSnapshot;
}

export type ScreenSharingAuxiliaryAction =
  | { readonly type: "start" | "stop" | "refresh-sources" | "clear-annotations" | "retry-pointers" }
  | { readonly type: "select-source"; readonly sourceId: number }
  | { readonly type: "set-pointers-enabled"; readonly enabled: boolean }
  | { readonly type: "set-interval"; readonly intervalSeconds: number };

export interface RoutedAuxiliaryAction {
  readonly revision: string;
  readonly pointerRevision: string;
  readonly action: ScreenSharingAuxiliaryAction;
}

export function isPointerSettingsAction(action: ScreenSharingAuxiliaryAction): boolean {
  return action.type === "set-pointers-enabled" || action.type === "retry-pointers";
}

export interface ScreenSharingAuxiliaryModel {
  readonly ownerKey: string;
  readonly available: boolean;
  readonly active: boolean;
  readonly busy: boolean;
  readonly pointersEnabled: boolean;
  readonly pointersReady: boolean;
  readonly sources: readonly { readonly id: number; readonly name: string }[];
  readonly sourceId: number | null;
  readonly intervalSeconds: number;
  readonly error?: string;
  readonly lastObservedAt?: number;
  readonly language: string;
  readonly start: () => Promise<void>;
  readonly stop: () => void;
  readonly refreshSources: () => Promise<void>;
  readonly clearAnnotations: () => Promise<void>;
  readonly retryPointers: () => Promise<void>;
  readonly setPointersEnabled: (enabled: boolean) => Promise<void>;
  readonly setSourceId: (id: number) => void;
  readonly setIntervalSeconds: (seconds: number) => void;
}

/** Copy known public fields only. In particular, errors may contain provider details. */
export function createScreenSharingSnapshot(
  model: ScreenSharingAuxiliaryModel,
  revision: string,
  pointerRevision = revision,
): ScreenSharingSnapshot {
  return {
    revision,
    pointerRevision,
    available: model.available,
    active: model.active,
    busy: model.busy,
    pointersEnabled: model.pointersEnabled,
    pointersReady: model.pointersReady,
    sources: model.sources.slice(0, 64).map(({ id, name }) => ({ id, name: name.slice(0, 200) })),
    sourceId: model.sourceId,
    intervalSeconds: model.intervalSeconds,
    hasError: Boolean(model.error),
    lastObservedAt: model.lastObservedAt ?? null,
    language: model.language.startsWith("ja") ? "ja" : "en",
  };
}

interface HostTransport {
  listenAction: (callback: (request: RoutedAuxiliaryAction) => void) => Promise<() => void>;
  publish: (snapshot: ScreenSharingSnapshot) => Promise<void>;
  open: () => Promise<void>;
  revision: () => string;
}

const nativeHostTransport: HostTransport = {
  listenAction: (callback) =>
    getCurrentWindow().listen<RoutedAuxiliaryAction>(AUXILIARY_ACTION_EVENT, (event) =>
      callback(event.payload),
    ),
  publish: (snapshot) => invoke("auxiliary_window_publish", { snapshot }),
  open: () => invoke("auxiliary_window_open", { kind: "screen-sharing-controls" }),
  revision: () => crypto.randomUUID(),
};

/** One main-window owner: no capture, session, or image delivery is performed by this bridge. */
export class ScreenSharingAuxiliaryHost {
  private model: ScreenSharingAuxiliaryModel | null = null;
  private snapshot: ScreenSharingSnapshot | null = null;
  private signature = "";
  private pointerSignature = "";
  private disposed = false;
  private unlisten: (() => void) | null = null;
  private publishQueue: Promise<void> = Promise.resolve();
  private publishError: unknown = null;
  private readonly ready: Promise<void>;

  constructor(
    private readonly onError: (error: unknown) => void,
    private readonly transport: HostTransport = nativeHostTransport,
  ) {
    this.ready = transport
      .listenAction((request) => {
        void this.handleAction(request).catch(onError);
      })
      .then((unlisten) => {
        if (this.disposed) unlisten();
        else this.unlisten = unlisten;
      });
    void this.ready.catch(onError);
  }

  update(model: ScreenSharingAuxiliaryModel): void {
    if (this.disposed) return;
    this.model = model;
    const safeFields = createScreenSharingSnapshot(model, "");
    // The owner affects revisions without revealing an agent or thread identifier to other views.
    const signature = JSON.stringify([model.ownerKey, safeFields]);
    if (signature === this.signature) return;
    this.signature = signature;
    const revision = this.transport.revision();
    const pointerSignature = JSON.stringify([
      model.ownerKey,
      model.pointersEnabled,
      model.pointersReady,
    ]);
    const pointerRevision =
      pointerSignature === this.pointerSignature && this.snapshot
        ? this.snapshot.pointerRevision
        : revision;
    this.pointerSignature = pointerSignature;
    const snapshot = { ...safeFields, revision, pointerRevision };
    this.snapshot = snapshot;
    this.publishQueue = this.publishQueue
      .then(async () => {
        if (this.disposed) return;
        await this.transport.publish(snapshot);
        this.publishError = null;
      })
      .catch((error: unknown) => {
        this.publishError = error;
        this.onError(error);
      });
  }

  async open(): Promise<void> {
    await this.ready;
    await this.publishQueue;
    if (this.disposed || !this.snapshot) return;
    if (this.publishError) throw this.publishError;
    await this.transport.open();
  }

  /** Check again in the owner: a newer React state may precede its native publication. */
  async handleAction(request: RoutedAuxiliaryAction): Promise<boolean> {
    const model = this.model;
    const snapshot = this.snapshot;
    if (this.disposed || !model || !snapshot) return false;
    const { action } = request;
    if (
      isPointerSettingsAction(action)
        ? request.pointerRevision !== snapshot.pointerRevision
        : request.revision !== snapshot.revision
    )
      return false;
    switch (action.type) {
      case "start":
        if (
          !model.available ||
          !model.pointersReady ||
          model.active ||
          model.busy ||
          !model.sources.some((source) => source.id === model.sourceId)
        )
          return false;
        await model.start();
        break;
      case "stop":
        model.stop();
        break;
      case "refresh-sources":
        if (model.active || model.busy) return false;
        await model.refreshSources();
        break;
      case "clear-annotations":
        await model.clearAnnotations();
        break;
      case "retry-pointers":
        if (model.pointersReady || !model.error) return false;
        await model.retryPointers();
        break;
      case "set-pointers-enabled":
        if (!model.pointersReady || typeof action.enabled !== "boolean") return false;
        await model.setPointersEnabled(action.enabled);
        break;
      case "select-source":
        if (
          model.active ||
          model.busy ||
          !model.sources.some((source) => source.id === action.sourceId)
        )
          return false;
        model.setSourceId(action.sourceId);
        break;
      case "set-interval":
        if (
          !Number.isInteger(action.intervalSeconds) ||
          action.intervalSeconds < 5 ||
          action.intervalSeconds > 60
        )
          return false;
        model.setIntervalSeconds(action.intervalSeconds);
        break;
    }
    return true;
  }

  dispose(): void {
    this.disposed = true;
    this.model = null;
    this.snapshot = null;
    this.unlisten?.();
    this.unlisten = null;
  }
}

export function readAuxiliarySnapshot(): Promise<PublishedAuxiliarySnapshot | null> {
  return invoke("auxiliary_window_snapshot");
}

export function requestAuxiliaryAction(
  version: number,
  action: ScreenSharingAuxiliaryAction,
  pointerRevision?: string,
): Promise<void> {
  return invoke("auxiliary_window_request_action", {
    request: { version, action, ...(pointerRevision === undefined ? {} : { pointerRevision }) },
  });
}

export function listenAuxiliarySnapshot(
  callback: (snapshot: PublishedAuxiliarySnapshot) => void,
): Promise<() => void> {
  return getCurrentWindow().listen<PublishedAuxiliarySnapshot>(AUXILIARY_STATE_EVENT, (event) =>
    callback(event.payload),
  );
}

/** A late initial read must not replace an event that already delivered newer state. */
export function latestAuxiliarySnapshot(
  current: PublishedAuxiliarySnapshot | null,
  incoming: PublishedAuxiliarySnapshot,
): PublishedAuxiliarySnapshot {
  return current && current.version > incoming.version ? current : incoming;
}

File: src-tauri/src/auxiliary_windows.rs
//! Host-owned auxiliary UI. A window kind selects a bundled view; callers never supply a URL.
//! Only the main view publishes state, and auxiliary actions go back to that same owner.

use serde::{Deserialize, Serialize};
use std::sync::Mutex;
use tauri::{AppHandle, Emitter, Manager, State, WebviewUrl, WebviewWindow, WebviewWindowBuilder};

const MAIN_LABEL: &str = "main";
const CONTROLS_LABEL: &str = "auxiliary-screen-sharing-controls";
const STATE_EVENT: &str = "auxiliary-window-state";
const ACTION_EVENT: &str = "auxiliary-window-action";

#[derive(Clone, Copy, Debug, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum AuxiliaryWindowKind {
    ScreenSharingControls,
}

impl AuxiliaryWindowKind {
    fn label(self) -> &'static str {
        match self {
            Self::ScreenSharingControls => CONTROLS_LABEL,
        }
    }

    fn route(self) -> &'static str {
        match self {
            Self::ScreenSharingControls => "screen-sharing-controls",
        }
    }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct SharedDisplay {
    id: u32,
    name: String,
}

#[derive(Clone, Copy, Debug, Default, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum SharingSourceKind {
    #[default]
    Screen,
    Camera,
}

/// Deliberately excludes image data, agent/thread identifiers, arbitrary error text, and credentials.
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ScreenSharingSnapshot {
    revision: String,
    pointer_revision: String,
    available: bool,
    active: bool,
    busy: bool,
    pointers_enabled: bool,
    pointers_ready: bool,
    sources: Vec<SharedDisplay>,
    #[serde(default)]
    source_kind: SharingSourceKind,
    source_id: Option<u32>,
    interval_seconds: u8,
    has_error: bool,
    last_observed_at: Option<u64>,
    language: String,
}

impl ScreenSharingSnapshot {
    fn validate(&self) -> Result<(), String> {
        if self.revision.is_empty()
            || self.revision.len() > 80
            || self.pointer_revision.is_empty()
            || self.pointer_revision.len() > 80
        {
            return Err("Invalid auxiliary state revision".into());
        }
        if !(20..=60).contains(&self.interval_seconds) {
            return Err("Viewing interval must be between 20 and 60 seconds".into());
        }
        if self.sources.len() > 64 || self.sources.iter().any(|source| source.name.len() > 800) {
            return Err("Invalid display list".into());
        }
        if self.language != "en" && self.language != "ja" {
            return Err("Unsupported auxiliary language".into());
        }
        Ok(())
    }
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(tag = "type", rename_all = "kebab-case", deny_unknown_fields)]
pub enum ScreenSharingAction {
    Start,
    Stop,
    RefreshSources,
    ClearAnnotations,
    RetryPointers,
    SetPointersEnabled {
        enabled: bool,
    },
    SelectSourceKind {
        #[serde(rename = "sourceKind")]
        source_kind: SharingSourceKind,
    },
    SelectSource {
        #[serde(rename = "sourceId")]
        source_id: u32,
    },
    SetInterval {
        #[serde(rename = "intervalSeconds")]
        interval_seconds: u8,
    },
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AuxiliaryActionRequest {
    version: u64,
    pointer_revision: Option<String>,
    action: ScreenSharingAction,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PublishedSnapshot {
    version: u64,
    snapshot: ScreenSharingSnapshot,
}

#[derive(Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct RoutedAction {
    revision: String,
    pointer_revision: String,
    action: ScreenSharingAction,
}

#[derive(Default)]
struct AuxiliaryState {
    version: u64,
    snapshot: Option<PublishedSnapshot>,
}

#[derive(Default)]
pub struct AuxiliaryWindowsState(Mutex<AuxiliaryState>);

fn require_label(actual: &str, required: &str) -> Result<(), String> {
    if actual == required {
        Ok(())
    } else {
        Err("This window cannot perform that auxiliary operation".into())
    }
}

fn is_allowed_navigation(url: &tauri::Url, main_url: &tauri::Url, query: &str) -> bool {
    // url::Origin is opaque for tauri:// on macOS, so compare its actual authority instead.
    url.scheme() == main_url.scheme()
        && url.host_str() == main_url.host_str()
        && url.port_or_known_default() == main_url.port_or_known_default()
        && matches!(url.path(), "/" | "/index.html")
        && url.query() == Some(query)
        && url.username().is_empty()
        && url.password().is_none()
}

fn validate_action(
    published: &PublishedSnapshot,
    request: &AuxiliaryActionRequest,
) -> Result<(), String> {
    let pointer_setting = matches!(
        &request.action,
        ScreenSharingAction::SetPointersEnabled { .. } | ScreenSharingAction::RetryPointers
    );
    let current_revision = if pointer_setting {
        request.pointer_revision.as_deref() == Some(published.snapshot.pointer_revision.as_str())
    } else {
        published.version == request.version
    };
    if !current_revision {
        return Err("Sharing settings changed. Please try again.".into());
    }
    let snapshot = &published.snapshot;
    match &request.action {
        ScreenSharingAction::Start
            if !snapshot.available
                || (snapshot.source_kind != SharingSourceKind::Camera
                    && !snapshot.pointers_ready)
                || snapshot.active
                || snapshot.busy
                || !snapshot
                    .sources
                    .iter()
                    .any(|source| Some(source.id) == snapshot.source_id) =>
        {
            Err("Screen sharing is not ready to start".into())
        }
        ScreenSharingAction::SetPointersEnabled { .. }
        | ScreenSharingAction::RetryPointers
        | ScreenSharingAction::ClearAnnotations
            if snapshot.source_kind == SharingSourceKind::Camera =>
        {
            Err("Desktop pointers are not available for camera sharing".into())
        }
        ScreenSharingAction::SetPointersEnabled { .. } if !snapshot.pointers_ready => {
            Err("Screen pointer settings are not ready".into())
        }
        ScreenSharingAction::RetryPointers if snapshot.pointers_ready || !snapshot.has_error => {
            Err("Screen pointer settings do not need initialization".into())
        }
        ScreenSharingAction::SelectSource { source_id }
            if snapshot.active
                || snapshot.busy
                || !snapshot
                    .sources
                    .iter()
                    .any(|source| source.id == *source_id) =>
        {
            Err("Stop sharing before selecting an available display".into())
        }
        ScreenSharingAction::RefreshSources if snapshot.active || snapshot.busy => {
            Err("Stop sharing before refreshing displays".into())
        }
        ScreenSharingAction::SetInterval { interval_seconds }
            if !(20..=60).contains(interval_seconds) =>
        {
            Err("Viewing interval must be between 20 and 60 seconds".into())
        }
        _ => Ok(()),
    }
}

/// Explicitly opening controls is the only path that changes auxiliary focus.
#[tauri::command]
pub async fn auxiliary_window_open(
    app: AppHandle,
    window: WebviewWindow,
    kind: AuxiliaryWindowKind,
) -> Result<(), String> {
    require_label(window.label(), MAIN_LABEL)?;
    if let Some(existing) = app.get_webview_window(kind.label()) {
        existing.unminimize().map_err(|error| error.to_string())?;
        existing.show().map_err(|error| error.to_string())?;
        return existing.set_focus().map_err(|error| error.to_string());
    }
    let main_url = window.url().map_err(|error| error.to_string())?;
    let query = format!("auxiliary={}", kind.route());
    let resource = format!("index.html?{query}");
    WebviewWindowBuilder::new(&app, kind.label(), WebviewUrl::App(resource.into()))
        .title("Screen sharing — Yorishiro")
        .inner_size(360.0, 530.0)
        .min_inner_size(320.0, 400.0)
        .resizable(true)
        .always_on_top(true)
        .focused(true)
        .skip_taskbar(true)
        .disable_drag_drop_handler()
        .on_navigation(move |url| is_allowed_navigation(url, &main_url, &query))
        .on_new_window(|_, _| tauri::webview::NewWindowResponse::Deny)
        .build()
        .map_err(|error| error.to_string())?;
    Ok(())
}

#[tauri::command]
pub fn auxiliary_window_publish(
    app: AppHandle,
    window: WebviewWindow,
    state: State<'_, AuxiliaryWindowsState>,
    snapshot: ScreenSharingSnapshot,
) -> Result<(), String> {
    require_label(window.label(), MAIN_LABEL)?;
    snapshot.validate()?;
    let published = {
        let mut state = state.0.lock().map_err(|error| error.to_string())?;
        state.version = state.version.saturating_add(1);
        let published = PublishedSnapshot {
            version: state.version,
            snapshot,
        };
        state.snapshot = Some(published.clone());
        published
    };
    // Never show or focus a window as a side effect of a state refresh or sharing stop.
    app.emit_to(CONTROLS_LABEL, STATE_EVENT, published)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub fn auxiliary_window_snapshot(
    window: WebviewWindow,
    state: State<'_, AuxiliaryWindowsState>,
) -> Result<Option<PublishedSnapshot>, String> {
    require_label(window.label(), CONTROLS_LABEL)?;
    Ok(state
        .0
        .lock()
        .map_err(|error| error.to_string())?
        .snapshot
        .clone())
}

#[tauri::command]
pub fn auxiliary_window_request_action(
    app: AppHandle,
    window: WebviewWindow,
    state: State<'_, AuxiliaryWindowsState>,
    request: AuxiliaryActionRequest,
) -> Result<(), String> {
    require_label(window.label(), CONTROLS_LABEL)?;
    let routed = {
        let state = state.0.lock().map_err(|error| error.to_string())?;
        let published = state
            .snapshot
            .as_ref()
            .ok_or("The main window is not ready")?;
        validate_action(published, &request)?;
        RoutedAction {
            revision: published.snapshot.revision.clone(),
            pointer_revision: published.snapshot.pointer_revision.clone(),
            action: request.action,
        }
    };
    if app.get_webview_window(MAIN_LABEL).is_none() {
        return Err("The main window is closed".into());
    }
    app.emit_to(MAIN_LABEL, ACTION_EVENT, routed)
        .map_err(|error| error.to_string())
}

/// Called when the owning main window is destroyed, including native close and app shutdown.
pub fn close_owned_windows(app: &AppHandle) {
    if let Some(window) = app.get_webview_window(CONTROLS_LABEL) {
        let _ = window.destroy();
    }
    if let Some(state) = app.try_state::<AuxiliaryWindowsState>() {
        if let Ok(mut state) = state.0.lock() {
            state.snapshot = None;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn published() -> PublishedSnapshot {
        PublishedSnapshot {
            version: 7,
            snapshot: ScreenSharingSnapshot {
                revision: "main-owner-revision".into(),
                pointer_revision: "pointer-owner-revision".into(),
                available: true,
                active: false,
                busy: false,
                pointers_enabled: true,
                pointers_ready: true,
                sources: vec![SharedDisplay {
                    id: 12,
                    name: "Display 1".into(),
                }],
                source_kind: SharingSourceKind::Screen,
                source_id: Some(12),
                interval_seconds: 30,
                has_error: false,
                last_observed_at: None,
                language: "ja".into(),
            },
        }
    }

    #[test]
    fn only_the_intended_window_can_publish_or_request() {
        assert!(require_label(MAIN_LABEL, MAIN_LABEL).is_ok());
        assert!(require_label(CONTROLS_LABEL, MAIN_LABEL).is_err());
        assert!(require_label(MAIN_LABEL, CONTROLS_LABEL).is_err());
        assert!(require_label("untrusted", CONTROLS_LABEL).is_err());
    }

    #[test]
    fn rejects_unknown_window_kinds_and_image_or_credential_fields() {
        assert!(serde_json::from_str::<AuxiliaryWindowKind>("\"https://example.com\"").is_err());
        let mut json = serde_json::to_value(published().snapshot).unwrap();
        for field in ["imageDataUrl", "token", "ownerKey", "error"] {
            json[field] = serde_json::json!("must not cross this bridge");
            assert!(serde_json::from_value::<ScreenSharingSnapshot>(json.clone()).is_err());
            json.as_object_mut().unwrap().remove(field);
        }
    }

    #[test]
    fn navigation_accepts_the_bundled_view_on_tauri_and_dev_origins_only() {
        let query = "auxiliary=screen-sharing-controls";
        for origin in [
            "tauri://localhost",
            "http://localhost:1430",
            "http://tauri.localhost",
        ] {
            let main = tauri::Url::parse(&format!("{origin}/")).unwrap();
            let bundled = tauri::Url::parse(&format!("{origin}/index.html?{query}")).unwrap();
            assert!(is_allowed_navigation(&bundled, &main, query));
            for destination in [
                format!("https://example.com/index.html?{query}"),
                format!("{origin}/index.html?auxiliary=unknown"),
                format!("{origin}/other.html?{query}"),
            ] {
                assert!(!is_allowed_navigation(
                    &tauri::Url::parse(&destination).unwrap(),
                    &main,
                    query
                ));
            }
        }
    }

    #[test]
    fn stale_controls_cannot_start_or_stop_a_replacement_share() {
        for action in [
            ScreenSharingAction::Start,
            ScreenSharingAction::Stop,
            ScreenSharingAction::RetryPointers,
        ] {
            assert!(validate_action(
                &published(),
                &AuxiliaryActionRequest {
                    version: 6,
                    pointer_revision: None,
                    action
                }
            )
            .is_err());
        }
    }

    #[test]
    fn validates_source_interval_and_availability() {
        let mut state = published();
        let request = |action| AuxiliaryActionRequest {
            version: 7,
            pointer_revision: None,
            action,
        };
        assert!(validate_action(&state, &request(ScreenSharingAction::Start)).is_ok());
        state.snapshot.available = false;
        assert!(validate_action(&state, &request(ScreenSharingAction::Start)).is_err());
        assert!(validate_action(
            &state,
            &request(ScreenSharingAction::SelectSource { source_id: 99 })
        )
        .is_err());
        for interval_seconds in [5, 19, 61] {
            assert!(validate_action(
                &state,
                &request(ScreenSharingAction::SetInterval { interval_seconds })
            )
            .is_err());
            state.snapshot.interval_seconds = interval_seconds;
            assert!(state.snapshot.validate().is_err());
        }
        for interval_seconds in [20, 30, 60] {
            assert!(validate_action(
                &state,
                &request(ScreenSharingAction::SetInterval { interval_seconds })
            )
            .is_ok());
            state.snapshot.interval_seconds = interval_seconds;
            assert!(state.snapshot.validate().is_ok());
        }
        state.snapshot.active = true;
        assert!(validate_action(
            &state,
            &request(ScreenSharingAction::SelectSource { source_id: 12 })
        )
        .is_err());
        assert!(validate_action(&state, &request(ScreenSharingAction::RefreshSources)).is_err());
        assert!(validate_action(&state, &request(ScreenSharingAction::Stop)).is_ok());
        assert!(validate_action(&state, &request(ScreenSharingAction::ClearAnnotations)).is_ok());
    }

    #[test]
    fn pointer_toggle_is_independent_of_capture_but_rejects_stale_or_unready_controls() {
        let mut state = published();
        state.snapshot.available = false;
        state.snapshot.busy = true;
        let request = |enabled| AuxiliaryActionRequest {
            version: state.version,
            pointer_revision: Some(state.snapshot.pointer_revision.clone()),
            action: ScreenSharingAction::SetPointersEnabled { enabled },
        };
        for enabled in [false, true] {
            assert!(validate_action(&state, &request(enabled)).is_ok());
        }
        let mut stale = request(false);
        stale.version -= 1;
        assert!(validate_action(&state, &stale).is_ok());
        stale.pointer_revision = Some("old-owner-or-setting".into());
        assert!(validate_action(&state, &stale).is_err());
        state.snapshot.pointers_ready = false;
        assert!(validate_action(&state, &request(false)).is_err());
        assert!(
            serde_json::from_value::<ScreenSharingAction>(serde_json::json!({
                "type": "set-pointers-enabled", "enabled": "true"
            }))
            .is_err()
        );
    }

    #[test]
    fn capture_updates_preserve_pointer_actions_but_owner_or_pointer_changes_reject_them() {
        let mut state = published();
        let off = AuxiliaryActionRequest {
            version: state.version,
            pointer_revision: Some(state.snapshot.pointer_revision.clone()),
            action: ScreenSharingAction::SetPointersEnabled { enabled: false },
        };
        state.version += 1;
        state.snapshot.revision = "capture-finished-revision".into();
        state.snapshot.last_observed_at = Some(1000);
        assert!(validate_action(&state, &off).is_ok());
        state.snapshot.pointer_revision = "new-owner-or-pointer-setting".into();
        assert!(validate_action(&state, &off).is_err());
    }
}
